### ISSUE TITLE
Batch size changes inference

### ISSUE THREAD

**Issue by ryan-caesar-ramos** [tloen/alpaca-lora#378]
Issue should be reproducible with [this notebook](https://colab.research.google.com/drive/1_g6WSdsKluqljXoi_uhNy3-YJLBEAgmo?usp=sharing) (EDIT: updated link, previous link was made editable by anyone).

(rest is copied from [this issue](https://github.com/huggingface/transformers/issues/22861) but they can't help with `decapoda-research/llama-7b-hf` which I assume `alpaca-lora` is for)

Basically when I run `generate` with just one tokenized sequence, I get a certain result, and when I process the same sequence but inside a batch instead, the result changes. To make sure it wasn't any tokenization shenanigans, I tokenized the whole batch and took just the parts that related to the weird sequence (basically just `{k: v[1:] for k, v in tokenized_inputs.items()}` where the batch is just two items and the weird one is the second item).

For clarity, when I pass just

```python3
"""Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.

### Instruction:
Combine the question and answer into an image caption as succinctly as possible. Be sure to include the phrase "a photo of". Do not draw false conclusions.

### Input:
Is this a baseball game? no

### Response:
"""
```

to the generate function of a `LlamaForCausalLM` wrapped in a `PeftModel`, I get

```
This is not a baseball game.
```

However, if I put it in a two-item batch, the output for some reason is instead

```
A photo of people playing a game.
```

**Comment by lywinged** [comment]
Just reproduced  your code with "decapoda-research/llama-7b-hf",
There are two issues:
1. Instruction format should be:
`instruction = "Combine the question and answer into an image caption as succinctly as possible. Be sure to include the phrase a photo of. Do not draw false conclusions."`
2. need to use batch_encode before batch_decode: 
`tokenized_inputs = tokenizer.batch_encode_plus(prompts, padding=True, truncation=False,return_tensors='pt').to('cuda')
`

Then the output would be:
```
<unk>Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.

### Instruction:
Combine the question and answer into an image caption as succinctly as possible. Be sure to include the phrase a photo of. Do not draw false conclusions.

### Input:
Is this a baseball game? no

### Response:
A photo of a baseball game.
--------------------------------------------------------------------------------
<unk>Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.

### Instruction:
Combine the question and answer into an image caption as succinctly as possible. Be sure to include the phrase a photo of. Do not draw false conclusions.

### Input:
Is this a baseball game? no

### Response:
A photo of a baseball game.
```

**Comment by ElleLeonne** [comment]
Batch size should change inference, no? Batch size is a method to aggregate and smooth the model's training, in an attempt to speed up the processing of massive changes to the gradient by processing more information at once. 

A lower batch size will result in the model seeing more nuanced, piece-meal, local minima data, which should result in different inference.

**Comment by ryan-caesar-ramos** [comment]
@ElleLeonne I'm not training though; what I mean is that during inference, model outputs are changing depending on how many outputs are being generated at a time, which shouldn't be the case. It's like having an image classifier that can correctly classify an image, however once you classify the same image but within a batch it starts changing its predictions.

@lywinged Thanks, I'll look into this! Mind if I ask what the problem with the original instruction is? Are there issues with using quotes?

**Comment by ryan-caesar-ramos** [comment]
Hi @lywinged I don't think the suggestions solve the underlying issue. If I follow both of them, different outputs are still being produced on `inputs = ["What country's flag is stitched to the back of the jeans? England", "What country's flag is stitched to the back of the jeans? usa"]`.

Btw doesn't a tokenizer's `__call__` function just call `batch_encode_plus` anyway if you pass in a list of strings?

**Comment by lywinged** [comment]
> Hi @lywinged I don't think the suggestions solve the underlying issue. If I follow both of them, different outputs are still being produced on `inputs = ["What country's flag is stitched to the back of the jeans? England", "What country's flag is stitched to the back of the jeans? usa"]`.
> 
> Btw doesn't a tokenizer's `__call__` function just call `batch_encode_plus` anyway if you pass in a list of strings?

https://colab.research.google.com/drive/1mBSGHmfZqb5MaFDaYfbNq5CedrBjEJb3?usp=share_link

**Comment by ryan-caesar-ramos** [comment]
Thanks, though the suggested changes still don't fix the underlying problem I think. Like I mentioned here

> Hi @lywinged I don't think the suggestions solve the underlying issue. If I follow both of them, different outputs are still being produced on `inputs = ["What country's flag is stitched to the back of the jeans? England", "What country's flag is stitched to the back of the jeans? usa"]`.
> 
> Btw doesn't a tokenizer's `__call__` function just call `batch_encode_plus` anyway if you pass in a list of strings?

you can change the inputs and get the same issue of "changing batch size during inference changes the generated output for some reason".

https://colab.research.google.com/drive/1AAk5uxLjdkmXDAJRocGul5eyucSaoS01?usp=sharing

**Comment by lywinged** [comment]
> Thanks, though the suggested changes still don't fix the underlying problem I think. Like I mentioned here
> 
> > Hi @lywinged I don't think the suggestions solve the underlying issue. If I follow both of them, different outputs are still being produced on `inputs = ["What country's flag is stitched to the back of the jeans? England", "What country's flag is stitched to the back of the jeans? usa"]`.
> > Btw doesn't a tokenizer's `__call__` function just call `batch_encode_plus` anyway if you pass in a list of strings?
> 
> you can change the inputs and get the same issue of "changing batch size during inference changes the generated output for some reason".
> 
> https://colab.research.google.com/drive/1AAk5uxLjdkmXDAJRocGul5eyucSaoS01?usp=sharing

```
print(tokenizer.batch_decode(full_output, skip_special_tokens=False)[0])
print('-' * 80)
print(tokenizer.batch_decode(full_output, skip_special_tokens=False)[1])
```

```
<unk><unk>Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.

### Instruction:
Combine the question and answer into an image caption as succinctly as possible. Be sure to include the phrase a photo of. Do not draw false conclusions.

### Input:
What country's flag is stitched to the back of the jeans? England

### Response:
A photo of England's flag stitched to the back of the jeans.
--------------------------------------------------------------------------------
<unk> Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.

### Instruction:
Combine the question and answer into an image caption as succinctly as possible. Be sure to include the phrase a photo of. Do not draw false conclusions.

### Input:
What country's flag is stitched to the back of the jeans? usa

### Response:
A photo of the USA flag stitched to the back of the jeans.<unk>
```

**Comment by ryan-caesar-ramos** [comment]
Sorry I'm not entirely sure what this code snippet

> ```
> print(tokenizer.batch_decode(full_output, skip_special_tokens=False)[0])
> print('-' * 80)
> print(tokenizer.batch_decode(full_output, skip_special_tokens=False)[1])
> ```

you shared is for. We're still getting different outputs for `tokenizer.batch_decode(full_output, skip_special_tokens=False)[1]` and `tokenizer.batch_decode(single_output, skip_special_tokens=False)[0]`, which is the problem.

**Comment by lywinged** [comment]
> Sorry I'm not entirely sure what this code snippet
> 
> > ```
> > print(tokenizer.batch_decode(full_output, skip_special_tokens=False)[0])
> > print('-' * 80)
> > print(tokenizer.batch_decode(full_output, skip_special_tokens=False)[1])
> > ```
> 
> you shared is for. We're still getting different outputs for `tokenizer.batch_decode(full_output, skip_special_tokens=False)[1]` and `tokenizer.batch_decode(single_output, skip_special_tokens=False)[0]`, which is the problem.

In fact, your question has nothing to do with alpaca-lora, that's basic skill issue, but share your notebook to me and make anyone editable.

**Comment by ryan-caesar-ramos** [comment]
Sure, here's the [link](https://colab.research.google.com/drive/1nAz2MUphzg5ifWW3CfRCJdYUgSYN5ynH?usp=sharing) to the original notebook, now editable.

Again, I'm not entirely sure what's happening here. Giving the same input to the model (with should have `do_sample=False` by default) yields different outputs depending on whether or not the input is within a batch with other inputs. Even with the suggested changes, this issue still arises with `inputs = ["What country's flag is stitched to the back of the jeans? England", "What country's flag is stitched to the back of the jeans? usa"]`. I'm not necessarily looking for "correct" answers (both of the answers I got for this input made sense) but the fact that it changes on the exact same input without sampling implies something is wrong.

**Comment by lywinged** [comment]
> Sure, here's the [link](https://colab.research.google.com/drive/1nAz2MUphzg5ifWW3CfRCJdYUgSYN5ynH?usp=sharing) to the original notebook, now editable.
> 
> Again, I'm not entirely sure what's happening here. Giving the same input to the model (with should have `do_sample=False` by default) yields different outputs depending on whether or not the input is within a batch with other inputs. Even with the suggested changes, this issue still arises with `inputs = ["What country's flag is stitched to the back of the jeans? England", "What country's flag is stitched to the back of the jeans? usa"]`. I'm not necessarily looking for "correct" answers (both of the answers I got for this input made sense) but the fact that it changes on the exact same input without sampling implies something is wrong.

It's done. Please check the notebook.

**Comment by ryan-caesar-ramos** [comment]
Unfortunately I can't test with `yahma/llama-7b-hf` because the shard sizes are too big for my runtime. Would you mind running this edited version of your last code cell?:

```
def generate_prompt(instruction, input):
    return f"Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.\n\n### Instruction:\n{instruction}\n\n### Input:\n{input}\n\n### Response:\n"

instruction_old = 'Combine the question and answer into an image caption as succinctly as possible. Be sure to include the phrase "a photo of". Do not draw false conclusions.'
instruction = "Combine the question and answer into an image caption as succinctly as possible. Be sure to include the phrase a photo of. Do not draw false conclusions."
inputs = ["What country's flag is stitched to the back of the jeans? England", "What country's flag is stitched to the back of the jeans? usa"]
prompts_old = [generate_prompt(instruction_old, inp) for inp in inputs]
print(prompts_old)
prompts = [generate_prompt(instruction, inp) for inp in inputs]
print(prompts)

tokenized_inputs = tokenizer(prompts, padding=True, truncation=True, return_tensors="pt").to('cuda')
tokenized_inputs_old = tokenizer(prompts_old, padding=True, truncation=True, return_tensors="pt").to('cuda')
print(tokenized_inputs_old)
print(tokenized_inputs)


generation_config = GenerationConfig(max_new_tokens=128)


#use old instruction with double quotation marks inside:
with torch.no_grad():
    full_output = model.generate(
        **tokenized_inputs_old,
        generation_config=generation_config
    )

    # same tokenized input, but we index just into the parts for the second input
    single_output = model.generate(
        **{k: v[1:] for k, v in tokenized_inputs_old.items()},
        generation_config=generation_config
    )

print(tokenizer.batch_decode(full_output, skip_special_tokens=False))
print('-' * 80)
print(tokenizer.batch_decode(full_output, skip_special_tokens=False)[-1]) 
print('-' * 80) 
print(tokenizer.batch_decode(single_output, skip_special_tokens=False)[0])
```

I just want to see if the issue persists with other inputs even with an updated tokenizer.

**Comment by lywinged** [comment]
> Unfortunately I can't test with `yahma/llama-7b-hf` because the shard sizes are too big for my runtime. Would you mind running this edited version of your last code cell?:
> 
> ```
> def generate_prompt(instruction, input):
>     return f"Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.\n\n### Instruction:\n{instruction}\n\n### Input:\n{input}\n\n### Response:\n"
> 
> instruction_old = 'Combine the question and answer into an image caption as succinctly as possible. Be sure to include the phrase "a photo of". Do not draw false conclusions.'
> instruction = "Combine the question and answer into an image caption as succinctly as possible. Be sure to include the phrase a photo of. Do not draw false conclusions."
> inputs = ["What country's flag is stitched to the back of the jeans? England", "What country's flag is stitched to the back of the jeans? usa"]
> prompts_old = [generate_prompt(instruction_old, inp) for inp in inputs]
> print(prompts_old)
> prompts = [generate_prompt(instruction, inp) for inp in inputs]
> print(prompts)
> 
> tokenized_inputs = tokenizer(prompts, padding=True, truncation=True, return_tensors="pt").to('cuda')
> tokenized_inputs_old = tokenizer(prompts_old, padding=True, truncation=True, return_tensors="pt").to('cuda')
> print(tokenized_inputs_old)
> print(tokenized_inputs)
> 
> 
> generation_config = GenerationConfig(max_new_tokens=128)
> 
> 
> #use old instruction with double quotation marks inside:
> with torch.no_grad():
>     full_output = model.generate(
>         **tokenized_inputs_old,
>         generation_config=generation_config
>     )
> 
>     # same tokenized input, but we index just into the parts for the second input
>     single_output = model.generate(
>         **{k: v[1:] for k, v in tokenized_inputs_old.items()},
>         generation_config=generation_config
>     )
> 
> print(tokenizer.batch_decode(full_output, skip_special_tokens=False))
> print('-' * 80)
> print(tokenizer.batch_decode(full_output, skip_special_tokens=False)[-1]) 
> print('-' * 80) 
> print(tokenizer.batch_decode(single_output, skip_special_tokens=False)[0])
> ```
> 
> I just want to see if the issue persists with other inputs even with an updated tokenizer.
```

['<unk><s> Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.\n\n### Instruction:\nCombine the question and answer into an image caption as succinctly as possible. Be sure to include the phrase "a photo of". Do not draw false conclusions.\n\n### Input:\nWhat country\'s flag is stitched to the back of the jeans? England\n\n### Response:\nA photo of England\'s flag stitched to the back of the jeans.</s>', '<s> Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.\n\n### Instruction:\nCombine the question and answer into an image caption as succinctly as possible. Be sure to include the phrase "a photo of". Do not draw false conclusions.\n\n### Input:\nWhat country\'s flag is stitched to the back of the jeans? usa\n\n### Response:\nA photo of the USA flag stitched to the back of the jeans.</s><unk>']
--------------------------------------------------------------------------------
<s> Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.

### Instruction:
Combine the question and answer into an image caption as succinctly as possible. Be sure to include the phrase "a photo of". Do not draw false conclusions.

### Input:
What country's flag is stitched to the back of the jeans? usa

### Response:
A photo of the USA flag stitched to the back of the jeans.</s><unk>
--------------------------------------------------------------------------------
<s> Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.

### Instruction:
Combine the question and answer into an image caption as succinctly as possible. Be sure to include the phrase "a photo of". Do not draw false conclusions.

### Input:
What country's flag is stitched to the back of the jeans? usa

### Response:
A photo of the USA flag stitched to the back of the jeans.</s>
```






I suggest to use cloud server, it's only $0.1 per hour for this small stuff.

**Comment by ryan-caesar-ramos** [comment]
Thanks! In that case looks like it was issues with the tokenizer. Still a little confused over how that's possible given that 1) the tokenization was the same across both batched and non-batched inference tests anyway and 2) `do_sample` was set to `False`, but the important thing is that the issue is sort of fixed. Thanks again!
